syz-manager/covfilter.go

```diff
@@ -136,7 +136,7 @@ func createCoverageBitmap(target *targets.Target, pcs map[uint32]uint32) []byte
 	// The file starts with two uint32: covFilterStart and covFilterSize,
 	// and a bitmap with size ((covFilterSize>>4) + 7)/8 bytes follow them.
 	// 8-bit = 1-byte, additional 1-byte to prevent overflow
-	data := make([]byte, 8+((size>>4)+7)/8)
+	data := make([]byte, 8+((size>>4)/8+1))
```

Comment:
The current logic looks correct to me. Go would panic on any out-of-bound accesses.
When does the out-of-bounds access happen?

Reply:
No, it's not correct, we faced the panic. 
size=37890432
pc=0x124f78a0
start=0x100d4f20

Reply:
Oh, I think I see the problem: we divide by 16 `size>>4` without rounding up.
The correct way to express this is then: `8+((size+15)/16+7)/8`.
Does it fix the crash with these values?

Reply:
so the size allocated for data[8:] is 296019 while calculation is 296019.875.
while ((pc-start)>>4)/8) is 296019.

Reply:
> Oh, I think I see the problem: we divide by 16 `size>>4` without rounding up.
> The correct way to express this is then: `8+((size+15)/16+7)/8`.
> Does it fix the crash with these values?

It will have the same panic as ((size+15)/16+7)/8 results in 296019.9921875 and still 296019.

Reply:
Mkay.
But adding 1 unconditionally still looks wrong, at least I can't understand why we are doing it. Which means me or somebody else can revert it back to +7/8 later.
Maybe we are off-by-one with size? Maybe size is max PC, which means size is actually size+1. It's like if max array index is 10, array size is 11.

Also is it possible to add a test for this?

Reply:
I'm also trying to see what the cov filter bitmap doing,
I put two values in test and it crashes too:

	pcs := map[uint32]uint32{
		0x810d4f20: 1,
		0x810d5020: 1,
	}

Reply:
> I put two values in test and it crashes too:

This is good, please add this to the test.

I am thinking maybe we need:

```
--- a/syz-manager/covfilter.go
+++ b/syz-manager/covfilter.go
@@ -166,7 +166,7 @@ func coverageFilterRegion(pcs map[uint32]uint32) (uint32, uint32) {
        // align
        start &= ^uint32(0xf)
        end = (end + 0xf) &^ uint32(0xf)
-       return start, end - start
+       return start, end - start + 1
 }
```

Reply:
> I'm also trying to see what the cov filter bitmap doing,

I allows syzkaller to concentrate on a specific part of the kernel only.
Say, even if you enable only networking syscalls, they will still trigger coverage also in mm/, fs/ and maybe other parts of the kernel. But if you are interested in net and net only, you may add `"files": ["net/"]` coverage filter, then syzkaller will ignore coverage in mm/fs and other dirs.

Reply:
no, it doesn't work, so are you trying to use each bit in bitmap to stands for one pc? If so, the current algo didn't distribute PCs well in these bitmap.

The new change "+       return start, end - start + 1" will failed at here
```
2021/03/23 19:14:11 coverage filter from 0x810d4f20 to 0x810d4fa1, size 0x81, pcs 2
129 1 1
0 bitmap=[0]
1 bitmap=[1]
--- FAIL: TestCreateBitmapArray (0.47s)
panic: runtime error: index out of range [1] with length 1 [recovered]
	panic: runtime error: index out of range [1] with length 1
```

Reply:
Actually my fix works:
You can see these two are not match:
 data := make([]byte, 8+((size>>4)+7)/8) // should be 8+((size>>4)/8 to match with below index.
pc = (pc - start) >> 4
bitmap[pc/8] |= (1 << (pc % 8))

The test code range to 300:
```
func TestCreateBitmapArray(t *testing.T) {

	target := targets.Get("test", "64")
	order := target.HostEndian
	var i uint32
	for i = 1; i < 300; i++ {
		pcs := map[uint32]uint32{
			0x810d4f20: 1,
		}
		pcs[0x810d4f20+i] = 1
		bitmap := createCoverageBitmap(target, pcs)
		start := order.Uint32(bitmap[0:])
		size := order.Uint32(bitmap[4:])
		fmt.Printf("i=%d size=%d, start=%d\n", i, size, start)
	}
}
```

Reply:
Are you sure? This does not crash for me:

```
--- a/syz-manager/covfilter.go
+++ b/syz-manager/covfilter.go
@@ -136,7 +136,7 @@ func createCoverageBitmap(target *targets.Target, pcs map[uint32]uint32) []byte
        // The file starts with two uint32: covFilterStart and covFilterSize,
        // and a bitmap with size ((covFilterSize>>4) + 7)/8 bytes follow them.
        // 8-bit = 1-byte, additional 1-byte to prevent overflow
-       data := make([]byte, 8+((size>>4)+7)/8)
+       data := make([]byte, 8+((size+15)/16+7)/8)
        order := binary.ByteOrder(binary.BigEndian)
        if target.LittleEndian {
                order = binary.LittleEndian
@@ -166,7 +166,7 @@ func coverageFilterRegion(pcs map[uint32]uint32) (uint32, uint32) {
        // align
        start &= ^uint32(0xf)
        end = (end + 0xf) &^ uint32(0xf)
-       return start, end - start
+       return start, end - start + 1
 }
 
 func compileRegexps(regexpStrings []string) ([]*regexp.Regexp, error) {
diff --git a/syz-manager/covfilter_test.go b/syz-manager/covfilter_test.go
index 0df988e561f2..736be2a59631 100644
--- a/syz-manager/covfilter_test.go
+++ b/syz-manager/covfilter_test.go
@@ -11,15 +11,15 @@ import (
 
 func TestCreateBitmap(t *testing.T) {
        pcs := map[uint32]uint32{
-               0x81000002: 1,
-               0x8120001d: 1,
+               0x810d4f20: 1,
+               0x810d5020: 1,
        }
```

Reply:
> Actually my fix works:

But it does not mean that it's correct ;)
`data := make([]byte, 1e6)` will also pass tests.

Reply:
ok, this patch works too. let me update

Reply:
true, still thinking how to do the test, loop until 0xffffffff seems not a good idea.

Reply:
I would loop for a range of 1-1000. It should be enough to cover 16-byte granules x 8-bit bytes.

Reply:
there is still bugs in your change.
1. end = (end + 0xf) &^ uint32(0xf) // overflow here too. And shouldn't do the 0xf operation as there is no such calculation in cov_filter.h.

2. And also, when PCmin ==0, PCmax>=0xfffffff1, it will crash too.

Ad for normal case, we need to guarantee every 128 address got one byte in bitmap. And (size>>4)/8+1 should be right as bitmap index starts from 0, while (size>>4)/8+1 gives that 1 room.

so a test for 0x100 should be enough:
```
	pcs = map[uint32]uint32{
		0x81000000: 1,
		0x81000100: 1,
	}
```
updated, please verify.

Reply:
Well, `(size>>4)/8+100` would be "correct" as well, but the problem is that for size=128 it creates 101-byte bitmap. And `(size>>4)/8+1` creates 2-byte bitmap. While we need only 1 byte. So it feel like we still don't understand what we are trying to do.

Reply:
> end = (end + 0xf) &^ uint32(0xf) // overflow here too. And shouldn't do the 0xf operation as there is no such calculation in cov_filter.h.

cov_filter.h does not have end at all, it already has size (cov_filter->pcsize) calculated by this covfilter.go.

Reply:
if size=128, required bitmap should be 2, right? ((maxPC-start)>>4)/8=1 in this case.
and (size>>4)/8+1 gives only 2.
That's correct,right?

Reply:
> > end = (end + 0xf) &^ uint32(0xf) // overflow here too. And shouldn't do the 0xf operation as there is no such calculation in cov_filter.h.
> 
> cov_filter.h does not have end at all, it already has size (cov_filter->pcsize) calculated by this covfilter.go.

oh, okay, so no just no need end+0xf, I think

Reply:
> if size=128, required bitmap should be 2, right? ((maxPC-start)>>4)/8=1 in this case.
> and (size>>4)/8+1 gives only 2.
> That's correct,right?

Why 2?
16 byte granules x 8-bits per byte = 128, which means for 128 PCs we need 1 byte

Reply:
0x81000000, 0x81000080, size is 128, actually address range is 129, so we need 2 bytes.

And I tested with this, succeed.
```
func TestBitmapArray(t *testing.T) {
	target := targets.Get("test", "64")
	for i := 0; i < 1000; i++ {
		pcs := map[uint32]uint32{
			0x81000002: 1,
		}
		pcs[0x81000002+uint32(i)] = 1
		createCoverageBitmap(target, pcs)
	}
	for i := 1000; i > 0; i-- {
		pcs := map[uint32]uint32{
			0xffffffff: 1,
		}
		pcs[0xffffffff-uint32(i)] = 1
		createCoverageBitmap(target, pcs)
	}
}
```

Reply:
I meant 0-128 has 129 addresses.

Reply:
By size I meant number of address. So consider range is 0-127. Your formula gives 2 bytes, while we need only 1.

Reply:
or not?...

Reply:
in python int((127>>4)/8+1)=1

And in go, both is 1.
```
123 1
124 1
125 1
126 1
127 1
128 2
129 2
130 2
```

Reply:
```
package main

import (
        "fmt"
)

func main() {
        for i := 0; i < 300;   {
                size := (i>>4)/8+1
                fmt.Printf("%d %d\n", i, size)
                i += 1
        }

}
```

Reply:
You code is equivalent provided that it uses effectively size-1:
https://play.golang.org/p/F8YClTM-_R8
I am just considerably confused by the fact that our size is effectively size - 1.